src/Operators/pointwisestencil.jl

```diff
@@ -0,0 +1,425 @@
+  - right_idx_matrix2 - idx < min(ubw1, j + lbw1 - 1),
+    min(ubw1, j + lbw1 - 1) <= min(ubw1, ubw1 + ubw2 - lbw2) =
+        ubw1 + min(0, ubw2 - lbw2) = ubw1, and
+    idx <= right_idx_matrix1                                                 ==>
+  - right_idx_matrix2 - idx < ubw1 and idx <= right_idx_matrix1              ==>
+  - right_idx_matrix2 - ubw1 < idx <= right_idx_matrix1                      ==>
+  - number of values of idx at the right boundary =
+        max(right_idx_matrix1 - (right_idx_matrix2 - ubw), 0)
+    The number of values of idx at the right boundary is 0 when
+        right_idx_matrix2 - ubw >= right_idx_matrix1.
+=#
+
+abstract type PointwiseStencilOperator <: FiniteDifferenceOperator end
+
+struct LeftStencilBoundary <: BoundaryCondition end
+struct RightStencilBoundary <: BoundaryCondition end
+
+has_boundary(
+    ::PointwiseStencilOperator,
+    ::LeftBoundaryWindow{name},
+) where {name} = true
+has_boundary(
+    ::PointwiseStencilOperator,
+    ::RightBoundaryWindow{name},
+) where {name} = true
+
+get_boundary(
+    ::PointwiseStencilOperator,
+    ::LeftBoundaryWindow{name},
+) where {name} = LeftStencilBoundary()
+get_boundary(
+    ::PointwiseStencilOperator,
+    ::RightBoundaryWindow{name},
+) where {name} = RightStencilBoundary()
+
+stencil_interior_width(::PointwiseStencilOperator, stencil, arg) =
+    ((0, 0), bandwidths(eltype(stencil)))
+
+function boundary_width(
+    ::PointwiseStencilOperator,
+    ::LeftStencilBoundary,
+    stencil,
+    arg,
+)
+    lbw = bandwidths(eltype(stencil))[1]
+    return max((left_idx(axes(arg)) - lbw) - left_idx(axes(stencil)), 0)
+end
+function boundary_width(
+    ::PointwiseStencilOperator,
+    ::RightStencilBoundary,
+    stencil,
+    arg,
+)
+    ubw = bandwidths(eltype(stencil))[2]
+    return max(right_idx(axes(stencil)) - (right_idx(axes(arg)) - ubw), 0)
+end
+
+##
+## ApplyStencil
+##
+
+# Matrix-Vector Product
+struct ApplyStencil <: PointwiseStencilOperator end
+
+# TODO: This is not correct for the same reason as the other 2-arg FD operators.
+return_eltype(::ApplyStencil, stencil, arg) = eltype(arg)
+
+return_space(::ApplyStencil, stencil_space, arg_space) = stencil_space
+
+function apply_stencil_at_idx(i_range, stencil, arg, loc, idx)
+    coefs = getidx(stencil, loc, idx)
+    lbw = bandwidths(eltype(stencil))[1]
+    i_func = i -> coefs[i - lbw + 1] ⊠ getidx(arg, loc, idx + i)
+    return ⊞(map(i_func, i_range)...)
+end
+
+function apply_stencil_at_idx_expr(i_range, stencil_type)
+    coefs_expr = :(coefs = getidx(stencil, loc, idx))
+    lbw = bandwidths(eltype(stencil_type))[1]
+    i_func = i -> :(coefs[$(i - lbw + 1)] ⊠ getidx(arg, loc, idx + $i))
+    return :(coefs_expr; return ⊞($(map(i_func, i_range)...)))
+end
+
+function stencil_interior(::ApplyStencil, loc, idx, stencil, arg)
+    lbw, ubw = bandwidths(eltype(stencil))
+    i_range = lbw:ubw
+    if @generated # TODO: Is this any better than the non-generated version?
```

Comment:
let's try to not use generated if it is not absolutely needed, can we test this again?

Reply:
I figured that these sorts of loops could be easily sped up, since the bandwidths are available at compile time. The one for ApplyStencil is pretty simple, but the one for ComposeStencils almost certainly can't be optimized away without generated functions. I didn't have time to do a proper comparison, though, so I could be wrong.

Reply:
Turns out the original implementation had a bug (eltype was not defined for field types, only for field instances). I just fixed the bug and ran a proper comparison. Without @generated:
```
julia> @timev compose.(stencil1, stencil2)
  0.054764 seconds (811.21 k allocations: 46.676 MiB)
elapsed time (ns): 54764000
bytes allocated:   48943552
pool allocs:       811210
malloc() calls:    1
```
With @generated:
```
julia> @timev compose.(stencil1, stencil2)
  0.008844 seconds (98.41 k allocations: 6.796 MiB)
elapsed time (ns): 8844500
bytes allocated:   7125952
pool allocs:       98410
malloc() calls:    1
```
I'd say an order-of-magnitude improvement is worth it, right?

Reply:
Yes absolutely.  Generated function have a cost so it is always good to check.  I don't think the way the non-generated function is written though is specializing on the bandwidth parameters, so that might be something to try.

Do you have to code used to generate the benchmark stencils?

Reply:
The test above was done by running `driver.jl`, then setting `stencil1 = W.∂𝕄ₜ∂ρ` and `stencil2 = W.∂ρₜ∂𝕄`.

When you say that it isn't specializing on the bandwidth parameters, do you mean that the compiler is treating the outputs of `bandwidths(eltype(stencil1))` and `bandwidths(eltype(stencil2))` as runtime values, rather than inferring them from the types of `stencil1` and `stencil2`? What's the best way to verify/fix this?

Reply:
you can specialize through there where clause, ex: 

```foo(stencil1::Setencil{S1BW1, S1B2}, stencil2::Stencil{S2B1, S2B2}) where {S1B1, S2B2, ...}```

see: https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing

you can check with `@code_lowered optimize=true ...`  you might need to stub out the broacast function into a toplevel function to do the later as broadcast get's lowered by the compiler into a series of calls.

Reply:
I tried your suggestion, but it did not improve the allocation/time of stencil composition at all. Unfortunately, `stencil_interior` is allocating a bunch of memory (which I checked using `perf/allocs.jl`) for stencil operations, which is definitely slowing down the implicit solver. Making `ComposeStencils` a generated function helps, but it does not eliminate the problem. If you have some time later, it might be useful for you to help me optimize it. I feel like I'm still not very certain about how to go about this sort of fine-tuning.

Reply:
Sounds good, we can look at it sometime early next week when you have some time.  I don't think it is a blocker but good to track down the issues.

Reply:
I tracked down the primary issue. I had a function that needed to find a tuple for each integer in some range. I implemented this by passing in a function that returned the tuple corresponding to each integer. However, the tuples returned by this function can all have different lengths, so the compiler was unable to infer the concrete output type of the passed function (it was just inferring `Tuple`). I have changed my code so that I now pass in a tuple of tuples instead of a function, and I index the tuple at a particular integer to get the tuple corresponding to that integer. Here are my new results:
Without `@generated`:
```
julia> @timev compose.(stencil1, stencil2)
  0.005454 seconds (84.00 k allocations: 6.135 MiB)
elapsed time (ns): 5454186
bytes allocated:   6433104
pool allocs:       84004
malloc() calls:    1
```
With `@generated`:
```
julia> @timev compose.(stencil1, stencil2)
  0.004084 seconds (84.00 k allocations: 6.135 MiB)
elapsed time (ns): 4084302
bytes allocated:   6433104
pool allocs:       84004
malloc() calls:    1
```
So, the generated version is still ~30% faster, but at least the non-generated version is not allocating unnecessary memory due to type inference failure. My stencil operations are still one of the dominant sources of allocations according to `perf/allocs.jl`, but now they are roughly as bad as other finite difference operations, rather than an order of magnitude worse.

Reply:
nice.  for measuring benchmarks with 1/1000th of a second difference you really need a statistical bench-marking test (ex. `@benchmark`) to make any diff conclusions, you cannot run one timing and give a conclusive answer.  It seems like they are pretty much eqv. so I would suggest dropping the `@generated` version.

Reply:
Here's a more rigorous test that does in-place operations and uses `@btime` (the first call is with `@generated`, and the second is without it):
```
julia> stencil1 = W.∂𝕄ₜ∂ρ; stencil2 = W.∂ρₜ∂𝕄; result = Operators.ComposeStencils().(stencil1, stencil2); compose!(result, stencil1, stencil2) = (result .= Operators.ComposeStencils().(stencil1, stencil2)); @btime compose!($result, $stencil1, $stencil2)
  3.034 ms (84000 allocations: 5.53 MiB)
ClimaCore.Operators.StencilCoefs{-1, 1, Tuple{Float64, Float64, Float64}}-valued Field:
  coefs: 
    1: [NaN, 0.0, -0.00223207, -0.00229962, -0.00239082, -0.0025035, -0.00262102, -0.00271167, -0.00278562, -0.00287097  …  0.0, -0.00223207, -0.00229962, -0.00239082, -0.0025035, -0.00262102, -0.00271167, -0.00278562, -0.00287097, 0.0]
    2: [-0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.0]
    3: [-0.0, 0.00119547, 0.00116286, 0.0011239, 0.00108114, 0.00104194, 0.00100713, 0.000970804, 0.00093571, 0.0  …  0.00119547, 0.00116286, 0.0011239, 0.00108114, 0.00104194, 0.00100713, 0.000970804, 0.00093571, 0.0, NaN]

julia> stencil1 = W.∂𝕄ₜ∂ρ; stencil2 = W.∂ρₜ∂𝕄; result = Operators.ComposeStencils().(stencil1, stencil2); compose!(result, stencil1, stencil2) = (result .= Operators.ComposeStencils().(stencil1, stencil2)); @btime compose!($result, $stencil1, $stencil2)
  4.309 ms (84000 allocations: 5.53 MiB)
ClimaCore.Operators.StencilCoefs{-1, 1, Tuple{Float64, Float64, Float64}}-valued Field:
  coefs: 
    1: [NaN, 0.0, -0.00223207, -0.00229962, -0.00239082, -0.0025035, -0.00262102, -0.00271167, -0.00278562, -0.00287097  …  0.0, -0.00223207, -0.00229962, -0.00239082, -0.0025035, -0.00262102, -0.00271167, -0.00278562, -0.00287097, 0.0]
    2: [-0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.0]
    3: [-0.0, 0.00119547, 0.00116286, 0.0011239, 0.00108114, 0.00104194, 0.00100713, 0.000970804, 0.00093571, 0.0  …  0.00119547, 0.00116286, 0.0011239, 0.00108114, 0.00104194, 0.00100713, 0.000970804, 0.00093571, 0.0, NaN]
```
So, the 30% difference seems to hold with better testing. Would you still say that 30% is not significant enough to merit a generated function?

Also, I've found that the other allocations are due to the same type instability problem, but at the boundaries rather than in the interior. Specifically, the tuple lengths at the boundaries depend on the values of `left_idx` and `right_idx` for the input spaces, and it seems like the compiler is unable to infer those values. It should be possible to make the inference succeed, since the number of levels of a space is a type parameter of the DataLayout, but I think I'll need your help to do that.